In [55]:
# HideCell
from IPython.display import HTML

HTML(
    """
    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap" rel="stylesheet">

    <link rel="stylesheet" href="custom.css">
    <script src="custom.js" defer></script>
    <style>
        div.jp-OutputPrompt.jp-OutputArea-prompt {
            display: none !important;
        }
    
        .jp-OutputPrompt {
            display: none !important;
        }

        .jp-Notebook-cell.celltag_remove_cell {
            display: none !important;
        }
    </style>

    """
)

In [ ]:
# HideCell
from IPython.display import HTML, Image, display


html_header = """

<header style="position: relative; overflow: hidden; background: #ffffff; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Inter, Roboto, Arial, sans-serif;">
  <div style="position: absolute; inset: 0; z-index: 0;" aria-hidden="true">
    <svg viewBox="0 0 1400 160" preserveAspectRatio="none" xmlns="http://www.w3.org/2000/svg" style="width:100%; height:100%; display:block;">
      <g stroke="#0c447c" stroke-opacity="0.06" stroke-width="1">
        <line x1="0" y1="40" x2="1400" y2="40"/><line x1="0" y1="80" x2="1400" y2="80"/><line x1="0" y1="120" x2="1400" y2="120"/>
      </g>
      <g fill="#7b828e" fill-opacity="0.28">
        <rect x="120" y="70" width="6" height="64"/><rect x="160" y="55" width="6" height="88"/><rect x="200" y="60" width="6" height="20"/><rect x="240" y="40" width="6" height="55"/><rect x="280" y="65" width="6" height="30"/><rect x="320" y="45" width="6" height="40"/><rect x="360" y="72" width="6" height="22"/><rect x="880" y="58" width="6" height="42"/><rect x="920" y="66" width="6" height="26"/><rect x="960" y="42" width="6" height="50"/><rect x="1000" y="60" width="6" height="32"/><rect x="1040" y="50" width="6" height="44"/><rect x="1080" y="68" width="6" height="24"/><rect x="1120" y="55" width="6" height="38"/>
      </g>
      <g stroke="#7b828e" stroke-opacity="0.28" stroke-width="1">
        <line x1="123" y1="60" x2="123" y2="110"/><line x1="163" y1="48" x2="163" y2="108"/><line x1="203" y1="52" x2="203" y2="86"/><line x1="243" y1="32" x2="243" y2="100"/><line x1="283" y1="58" x2="283" y2="100"/><line x1="323" y1="38" x2="323" y2="90"/><line x1="363" y1="65" x2="363" y2="100"/><line x1="883" y1="50" x2="883" y2="104"/><line x1="923" y1="58" x2="923" y2="96"/><line x1="963" y1="34" x2="963" y2="96"/><line x1="1003" y1="52" x2="1003" y2="96"/><line x1="1043" y1="42" x2="1043" y2="98"/><line x1="1083" y1="60" x2="1083" y2="96"/><line x1="1123" y1="46" x2="1123" y2="97"/>
      </g>
      <polyline fill="none" stroke="#5a6270" stroke-opacity="0.35" stroke-width="1.5"
        points="0,120 60,110 140,95 220,100 300,75 380,85 460,60 540,68 620,45 700,58 780,38 860,50 940,30 1020,42 1100,25 1180,35 1260,18 1400,28"/>
    </svg>
  </div>
  <div style="position: relative; z-index: 1; max-width: 1200px; margin: 0 auto; padding: 18px 24px; display: flex; align-items: center; justify-content: space-between; gap: 24px;">
    <div style="display: flex; align-items: center; gap: 12px;">
      <svg width="30" height="60" viewBox="0 0 30 30" xmlns="http://www.w3.org/2000/svg" role="img" aria-label="Database icon" style="flex-shrink:0;">
        <ellipse cx="15" cy="6" rx="11" ry="4" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <path d="M4 6 V15 C4 17.2 9 19 15 19 C21 19 26 17.2 26 15 V6" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <path d="M4 15 V24 C4 26.2 9 28 15 28 C21 28 26 26.2 26 24 V15" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <ellipse cx="15" cy="15" rx="11" ry="4" fill="none" stroke="#7b828e" stroke-width="1.4"/>
      </svg>
    </div>
  </div>
</header>
"""

display(HTML(html_header))


In [71]:
HTML("""
<script>
document.body.classList.add('preload');
</script>
""")

In [72]:
HTML("""
<script>
window.addEventListener('load', () => {
    document.body.classList.remove('preload');
});
</script>
""")

In [69]:
# HideCell
from IPython.display import HTML, Image, display
from datetime import date

today_str = date.today().strftime("%B %d, %Y")

html_coverpage_container = f"""
<div id="cover-page-container">
    <div id="title-container">
        <div id="title">Indicators</div>
        <div id="meta-data-container">
            <div>{today_str}</div>
            <div>by Aaron Hardy</div>
        </div>
    </div>
    <div id="title-bar"></div>
</div>
"""


display(HTML(html_coverpage_container))

# Summary

This project involved transforming raw fundamentals data into a structured set of financial ratios that can be used to reveal the underlying health, performance, and risk profile of a company. Raw accounting numbers alone rarely tell a clear story - ratios translate those numbers into standardized, comparable signals. By computing profitability, liquidity, solvency, efficiency, and valuation ratios, the project creates a foundation for deeper analytics, trend detection, and investment‑grade insights. The result is a unified dataset that supports downstream modeling, dashboards, and decision‑making.

# The Data
The raw data was acquired from [SimFin.com](https://www.simfin.com/en/fundamental-data-download/), which requires a free account to download. MySQL was used as the database management system. The raw data existed as CSVs. Each CSV was loaded into a MySQL table. The code used to load each table can be seen [here](load_data.html)

In [ ]:
SELECT * 
FROM industries 
ORDER BY IndustryId ASC 
LIMIT 10;

In [54]:
# HideCell
from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT *
FROM industries 
ORDER BY IndustryId ASC 
LIMIT 10;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
df = df.map(lambda x: x.strip() if isinstance(x, str) else x)
HTML(df.to_html(index=False))

IndustryId,Industry,Sector
100001,Industrial Products,Industrials
100002,Business Services,Industrials
100003,Engineering & Construction,Industrials
100004,Waste Management,Industrials
100005,Industrial Distribution,Industrials
100006,Airlines,Industrials
100007,Consulting & Outsourcing,Industrials
100008,Aerospace & Defense,Industrials
100009,Farm & Construction Machinery,Industrials
100010,Transportation & Logistics,Industrials


In [ ]:
SELECT * 
FROM us_balance_quarterly
LIMIT 10;

In [58]:
# HideCell
from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT * 
FROM us_balance_quarterly
LIMIT 10;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.to_html(index=False))

Ticker,SimFinId,Currency,FiscalYear,FiscalPeriod,ReportDate,PublishDate,RestatedDate,SharesBasic,SharesDiluted,CashAndEquiv,AccountsNotesReceivable,Inventories,TotalCurrentAssets,PropertyPlantEquipmentNet,LongTermInvestmentsReceivables,OtherLongTermAssets,TotalNoncurrentAssets,TotalAssets,PayablesAccruals,ShortTermDebt,TotalCurrentLiabilities,LongTermDebt,TotalNoncurrentLiabilities,TotalLiabilities,ShareCapitalAdditionalPaidIn,TreasuryStock,RetainedEarnings,TotalEquity,TotalLiabilitiesEquity
A,45846,USD,2020,Q3,2020-07-31,2020-09-01,2020-09-01,309000000,312000000,1358000000,930000000,746000000,3245000000,846000000,148000000,5307000000,6301000000,9546000000,311000000,40000000,1314000000,2283000000,3251000000,4565000000,5327000000,0,130000000,4981000000,9546000000
A,45846,USD,2020,Q4,2020-10-31,2020-12-18,2021-12-17,308000000,311000000,1441000000,1038000000,720000000,3415000000,845000000,158000000,5209000000,6212000000,9627000000,639000000,75000000,1467000000,2284000000,3287000000,4754000000,5314000000,0,81000000,4873000000,9627000000
A,45846,USD,2021,Q1,2021-01-31,2021-03-02,2021-03-02,306000000,309000000,1329000000,1087000000,755000000,3483000000,866000000,165000000,5160000000,6191000000,9674000000,656000000,314000000,1687000000,2185000000,3183000000,4870000000,5269000000,0,4000000,4804000000,9674000000
A,45846,USD,2021,Q2,2021-04-30,2021-06-01,2021-06-01,306000000,306000000,1380000000,1075000000,791000000,3514000000,884000000,188000000,5812000000,6884000000,10398000000,738000000,205000000,1758000000,2727000000,3830000000,5588000000,5274000000,0,-12000000,4810000000,10398000000
A,45846,USD,2021,Q3,2021-07-31,2021-09-01,2021-09-01,303000000,306000000,1428000000,1122000000,818000000,3632000000,905000000,204000000,5750000000,6859000000,10491000000,728000000,130000000,1724000000,2728000000,3821000000,5545000000,5310000000,0,90000000,4946000000,10491000000
A,45846,USD,2021,Q4,2021-10-31,2021-12-17,2022-12-21,305000000,307000000,1575000000,1172000000,830000000,3799000000,945000000,185000000,5776000000,6906000000,10705000000,774000000,0,1708000000,2729000000,3608000000,5316000000,5323000000,0,348000000,5389000000,10705000000
A,45846,USD,2022,Q1,2022-01-31,2022-03-03,2022-03-03,301000000,303000000,1158000000,1205000000,879000000,3474000000,974000000,191000000,5688000000,6853000000,10327000000,801000000,0,1584000000,2730000000,3589000000,5173000000,5293000000,0,159000000,5154000000,10327000000
A,45846,USD,2022,Q2,2022-04-30,2022-05-31,2022-05-31,299000000,301000000,1207000000,1237000000,937000000,3643000000,1010000000,190000000,5612000000,6812000000,10455000000,780000000,175000000,1813000000,2730000000,3520000000,5333000000,5295000000,0,160000000,5122000000,10455000000
A,45846,USD,2022,Q3,2022-07-31,2022-09-01,2023-03-03,298000000,299000000,1053000000,1405000000,1038000000,3778000000,1100000000,195000000,5459000000,6754000000,10532000000,909000000,36000000,1861000000,2733000000,3366000000,5227000000,5328000000,0,324000000,5305000000,10532000000
A,45846,USD,2022,Q4,2022-10-31,2022-12-21,2023-12-20,318000000,325000000,1053000000,1405000000,1038000000,3778000000,1100000000,195000000,5459000000,6754000000,10532000000,909000000,36000000,1861000000,2733000000,3366000000,5227000000,5328000000,0,324000000,5305000000,10532000000


The list of U.S. companies Excel file was inspected in Excel and the file contained rows of data split over several cells in an unpredictable way. In Excel, the CONCAT() function was used to convert all rows to a single concatenated string which was then split again using Data > Text to Columns tool.

Next, each SimFin CSV file was uploaded into a separate SQL database table. The example below shows how the table `us_companies` was uploaded to the MySQL database titled 'fundamentals_data'. All SimFin data files were uploaded similarly.

In [ ]:
DROP TABLE IF EXISTS us_companies;

-- Create the table structure
CREATE TABLE us_companies (
    Ticker VARCHAR(255),
    SimFinId INT,
    CompanyName VARCHAR(255),
    IndustryId INT,
    ISIN VARCHAR(255),
    EndOfFinancialYearMonth INT,
    NumberEmployees INT,
    BusinessSummary TEXT,
    Market VARCHAR(255),
    CIK INT,
    MainCurrency VARCHAR(255)
);

-- Upload the csv data
LOAD DATA LOCAL INFILE 'c:\\Users\\aaron\\Desktop\\Local Projects\\Portfolio Projects\\data\\SimFin\\us-companies\\us-companies.csv'
INTO TABLE us_companies
FIELDS TERMINATED BY ','
OPTIONALLY ENCLOSED BY '"'
LINES TERMINATED BY '\n'
IGNORE 1 LINES; -- skip header row


The company profile table (`us_companies`) had 37 rows where both the Ticker and Company name were NULL.

In [ ]:
SELECT * FROM us_companies;

UPDATE us_companies
SET `Company Name` = NULL, `Ticker` = NULL
WHERE `Company Name` = '' AND `Ticker` = '';

SELECT COUNT(*) 
FROM us_companies
WHERE (Ticker IS NULL AND `Company Name` IS NULL);


Next, we deleted from the table any rows that had a NULL value for `Ticker` and `CompanyName`. These rows had insufficient identifying information.

In [ ]:

DELETE FROM us_companies 
WHERE (Ticker IS NULL AND CompanyName IS NULL);

The data includes several financial statement tables. The range of dates for all balance sheet, income statement, and cash flow tables were confirmed below.

In [ ]:

SELECT 
    "us_balance_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_quarterly
UNION ALL
SELECT 
    "us_balance_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_annual
UNION ALL
SELECT 
    "us_balance_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_ttm
UNION ALL
SELECT 
    "us_income_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_quarterly
UNION ALL
SELECT 
    "us_income_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_annual
UNION ALL
SELECT 
    "us_income_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_ttm
UNION ALL
SELECT 
    "us_cashflow_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_quarterly
UNION ALL
SELECT 
    "us_cashflow_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_annual
UNION ALL
SELECT 
    "us_cashflow_ttm" AS table_name,        
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_ttm;





In [61]:
# HideCell
import json
import mysql.connector

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

cursor = conn.cursor()
cursor.execute(f"ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '{creds['password']}';")
cursor.execute("FLUSH PRIVILEGES;")

In [ ]:
# HideCell
from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT 
    "us_balance_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_quarterly
UNION ALL
SELECT 
    "us_balance_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_annual
UNION ALL
SELECT 
    "us_balance_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_ttm
UNION ALL
SELECT 
    "us_income_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_quarterly
UNION ALL
SELECT 
    "us_income_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_annual
UNION ALL
SELECT 
    "us_income_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_ttm
UNION ALL
SELECT 
    "us_cashflow_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_quarterly
UNION ALL
SELECT 
    "us_cashflow_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_annual
UNION ALL
SELECT 
    "us_cashflow_ttm" AS table_name,        
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_ttm;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.to_html(index=False))

table_name,min_date,max_date,years
us_balance_quarterly,2020-07-31,2025-05-31,4.8323
us_balance_annual,2020-07-31,2025-05-31,4.8323
us_balance_ttm,2020-07-31,2025-05-31,4.8323
us_income_quarterly,2020-07-31,2025-05-31,4.8323
us_income_annual,2020-07-31,2025-05-31,4.8323
us_income_ttm,2020-07-31,2025-05-31,4.8323
us_cashflow_quarterly,2020-07-31,2025-05-31,4.8323
us_cashflow_annual,2020-07-31,2025-05-31,4.8323
us_cashflow_ttm,2020-07-31,2025-05-31,4.8323


The data spans a period starting *July 31, 2020* and ending *May 31, 2025*, covering about 4.8 years.

Next, we looked at the minimum and maximum prices for both `Close` and `AdjClose`. 

In [ ]:
SELECT 
    MIN(Close),
    MAX(Close),
    MIN(AdjClose),
    MAX(AdjClose)
FROM us_shareprices_daily;

In [ ]:
# HideCell

from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT 
    MIN(Close),
    MAX(Close),
    MIN(AdjClose),
    MAX(AdjClose)
FROM us_shareprices_daily;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.to_html(index=False))

MIN(Close),MAX(Close),MIN(AdjClose),MAX(AdjClose)
0.0000,99999999.9900,0.0000,99999999.9900


The `Close` price ranged from \\$0 to \\$99,999,999.99. 
The `Adj. Close` price ranged from \\$0 to \\$9,984,240.00.

We removed tickers where a `Close` or `AdjClose` price for any day exceeded \\$100,000, since the common investor is not likely to have an investing budget exceeding \\$100,000 for any single ticker.

In [ ]:
SELECT DISTINCT Ticker
    FROM us_shareprices_daily
    WHERE Close > 100000 OR AdjClose > 100000;

In [ ]:
# HideCell

from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT DISTINCT Ticker
    FROM us_shareprices_daily 
    WHERE Close > 100000 OR AdjClose > 100000;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
csv_string = ", ".join(df["Ticker"].astype(str))
print (f"{len(df)} tickers:")
print(csv_string)



31 tickers:
ALLR, APDN, APVO, BCAN, BRK-A, CETX, CJET, CYN, EJH, ERNA, EZGO, FFIE, GDHG, HCTI, HLBZ, HOLO, HSDT, HUBC, JAGX, KXIN, MCMJ, MULN, PIXY, SHPH, TGL, TNXP, VCIG, VLCN, WHLR, WKHS, XXII


In [ ]:
DROP TABLE IF EXISTS us_shareprices_daily_filtered;

CREATE TABLE us_shareprices_daily_filtered AS
SELECT *
FROM us_shareprices_daily
WHERE
    Ticker NOT IN(
        SELECT DISTINCT
            Ticker
        FROM us_shareprices_daily
        WHERE
            Close > 100000
            OR AdjClose > 100000
    )

SELECT * FROM us_shareprices_daily_filtered LIMIT 10;

WITH full_table AS (
        SELECT *, ROW_NUMBER() OVER (
                PARTITION BY
                    Ticker
                ORDER BY Date DESC
            ) AS rn
        FROM us_shareprices_daily_filtered
)
SELECT *
FROM full_table
WHERE rn <= 3
LIMIT 9

In [ ]:
# HideCell

from IPython.display import HTML 
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}")


query = """
WITH full_table AS (
        SELECT *, ROW_NUMBER() OVER (
                PARTITION BY
                    Ticker
                ORDER BY Date DESC
            ) AS rn
        FROM us_shareprices_daily_filtered
)
SELECT *
FROM full_table
WHERE rn <= 3
LIMIT 9
"""

df = pd.read_sql(query, engine)
HTML(df.to_html(index=False))

# -- df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
# -- HTML(df.to_html(index=False))

Ticker,SimFinId,Date,Open,High,Low,Close,AdjClose,Volume,Dividend,SharesOutstanding,rn
,18728220,2025-06-20,12.19,13.37,11.49,12.17,12.17,167628,0.0,15674052,1
,18728220,2025-06-18,9.16,12.00,9.15,11.96,11.96,84360,0.0,15674052,2
,18728220,2025-06-17,9.01,10.14,9.01,9.84,9.84,85917,0.0,15674052,3
A,45846,2025-06-20,116.17,116.86,115.08,115.56,114.65,3043869,0.0,284064728,1
A,45846,2025-06-18,116.09,116.88,114.67,115.52,114.61,1560338,0.0,284064728,2
A,45846,2025-06-17,117.50,118.47,115.82,116.09,115.17,1577141,0.0,284064728,3
AA,367153,2025-06-20,28.53,28.61,27.83,27.86,27.60,7894057,0.0,258902338,1
AA,367153,2025-06-18,28.63,28.85,28.13,28.35,28.09,5634805,0.0,258902338,2
AA,367153,2025-06-17,29.05,29.49,28.61,28.72,28.46,5188049,0.0,258902338,3


Next, I aggregated the daily stock prices into monthly prices.

In [ ]:
DROP TABLE IF EXISTS us_shareprices_monthly;
-- When you drop the table, all indexes disappear automatically, so you must recreate them.

-- CREATE TABLE shareprices_monthly AS
-- SELECT *
-- FROM (
--     SELECT
--         Ticker,
--         SimFinId,
--         DATE_FORMAT(Date, '%Y-%m') AS Month,
--         Date,
--         Close,
--         AdjClose,
--         ROW_NUMBER() OVER (
--             PARTITION BY SimFinId, DATE_FORMAT(Date, '%Y-%m')
--             ORDER BY Date DESC
--         ) AS rn
--     FROM us_shareprices_daily_filtered
-- ) x
-- WHERE rn = 1;

SELECT * FROM us_shareprices_monthly LIMIT 10;

CREATE TABLE us_shareprices_monthly AS
SELECT
    Ticker,
    SimFinId,
    MAX(Date) AS Date,
    -- GROUP_CONCAT() → builds a list of Close prices
	-- ORDER BY Date DESC → newest price first
	-- SUBSTRING_INDEX(..., ',', 1) → pick the first price
    SUBSTRING_INDEX(
        GROUP_CONCAT(Close ORDER BY Date DESC),
        ',',
        1
    ) AS Close,
    SUBSTRING_INDEX(
        GROUP_CONCAT(AdjClose ORDER BY Date DESC),
        ',',
        1
    ) AS AdjClose
FROM us_shareprices_daily_filtered
GROUP BY
    Ticker,
    SimFinId,
    YEAR(Date),
    MONTH(Date);
    
    
CREATE INDEX idx_month_end_ticker_date
ON us_shareprices_monthly (SimFinId, Date);


# Ratios & Metrics

## Miscellaneous Metrics
Next, miscellaneous metrics such as effective tax rate, market cap, total debt, and free cash flow were computed and inserted into a new table, `metrics_misc`

In [ ]:
DROP TABLE IF EXISTS metrics_misc;

CREATE TABLE metrics_misc AS (
    WITH
        cte AS (
            SELECT
                isq.Ticker,
                isq.ReportDate,
                bsq.CashAndEquiv,
                CASE
                    WHEN isq.IncomeTaxExpenseBenefitNet >= 0 THEN NULL
                    WHEN isq.PretaxIncomeLoss <= 0 THEN NULL
                    ELSE -1 * isq.IncomeTaxExpenseBenefitNet / NULLIF(isq.PretaxIncomeLoss, 0)
                END AS EffectiveTaxRate,
                cfq.SharesBasic * (
                    SELECT pr.Close
                    FROM us_shareprices_daily pr
                    WHERE
                        pr.Ticker = isq.Ticker
                        AND pr.Date <= isq.ReportDate
                    ORDER BY pr.Date DESC
                    LIMIT 1
                ) AS MarketCap,
                bsq.ShortTermDebt + bsq.LongTermDebt AS TotalDebt,
                cfq.NetCashFromOperatingActivities + cfq.ChangeInFixedAssetsIntangibles AS FreeCashFlow
            FROM
                us_income_quarterly isq
                LEFT JOIN us_cashflow_quarterly cfq USING(Ticker, ReportDate)
                LEFT JOIN us_balance_quarterly bsq USING(Ticker, ReportDate)
        )
    SELECT cte.*, cte.MarketCap + cte.TotalDebt - cte.CashAndEquiv AS EnterpriseValue
    FROM cte
)

In [ ]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)
\
df = pd.read_sql("SELECT * FROM metrics_misc", engine)

# -- df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.tail().to_html(index=False))


Ticker,ReportDate,CashAndEquiv,EffectiveTaxRate,MarketCap,TotalDebt,FreeCashFlow,EnterpriseValue
ZYXI,2024-03-31,32861000.0,0.2308,NaN,57839000.0,1915000.0,NaN
ZYXI,2024-06-30,30896000.0,0.2464,NaN,58078000.0,997000.0,NaN
ZYXI,2024-09-30,37630000.0,0.2502,NaN,58320000.0,7002000.0,NaN
ZYXI,2024-12-31,39631000.0,NaN,NaN,58567000.0,2229000.0,NaN
ZYXI,2025-03-31,23852000.0,NaN,NaN,58818000.0,-10671000.0,NaN


## Growth Metrics

Growth rates - both year-on-year and quarter-on-quarter - are useful for trend analysis. The year-on-year (YoY) changes for various key series were computed and added to a `metrics_growth` table. 

In [ ]:
DROP TABLE IF EXISTS metrics_growth;

CREATE TABLE metrics_growth AS
SELECT
    isq.Ticker,
    isq.ReportDate,
    Revenue / NULLIF(LAG(Revenue, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS RevenueGrowthYOY,
    CASE
        WHEN GrossProfit < 0
        OR LAG(GrossProfit, 4) OVER(PARTITION BY Ticker ORDER BY ReportDate) <= 0 THEN NULL
        ELSE GrossProfit / NULLIF(LAG(GrossProfit, 4) OVER (
            PARTITION BY
                Ticker
            ORDER BY ReportDate
        ), 0) -1
    END AS GrossProfitGrowthYOY,
    CASE
        WHEN OperatingIncomeLoss < 0
        OR LAG(OperatingIncomeLoss, 4) OVER(PARTITION BY Ticker ORDER BY ReportDate) <= 0 THEN NULL
        ELSE OperatingIncomeLoss / NULLIF(LAG(OperatingIncomeLoss, 4) OVER (
            PARTITION BY
                Ticker
            ORDER BY ReportDate
        ), 0) -1 
    END AS OperatingIncomeGrowthYOY,
    CASE
        WHEN NetIncome < 0
        OR LAG(NetIncome, 4) OVER(PARTITION BY Ticker ORDER BY ReportDate) <= 0 THEN NULL
        ELSE NetIncome / NULLIF(LAG(NetIncome, 4) OVER (
            PARTITION BY
                Ticker
            ORDER BY ReportDate
        ), 0) -1
    END AS NetIncomeGrowthYOY,
    TotalAssets / NULLIF(LAG(TotalAssets, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS TotalAssetsGrowthYOY,
    TotalLiabilities / NULLIF(LAG(TotalLiabilities, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS TotalLiabilitiesGrowthYOY,
    CashAndEquiv / NULLIF(LAG(CashAndEquiv, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS CashAndEquivGrowthYOY,
    NetCashFromOperatingActivities / NULLIF(LAG(NetCashFromOperatingActivities, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS NetCashFromOperatingActivitiesYOY,
    NetCashFromInvestingActivities / NULLIF(LAG(NetCashFromInvestingActivities, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS NetCashFromInvestingActivitiesYOY,
    NetCashFromFinancingActivities / NULLIF(LAG(NetCashFromFinancingActivities, 4) OVER (
        PARTITION BY
            Ticker
        ORDER BY ReportDate
    ), 0) -1 AS NetCashFromFinancingActivitiesYOY 
FROM
    us_income_quarterly isq
    LEFT JOIN us_balance_quarterly bsq USING(Ticker, ReportDate)
    LEFT JOIN us_cashflow_quarterly cfq USING(Ticker, ReportDate)

In [ ]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)
\
df = pd.read_sql("SELECT * FROM metrics_growth", engine)

# -- df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.tail().to_html(index=False))


Ticker,ReportDate,RevenueGrowthYOY,GrossProfitGrowthYOY,OperatingIncomeGrowthYOY,NetIncomeGrowthYOY,TotalAssetsGrowthYOY,TotalLiabilitiesGrowthYOY,CashAndEquivGrowthYOY,NetCashFromOperatingActivitiesYOY,NetCashFromInvestingActivitiesYOY,NetCashFromFinancingActivitiesYOY
ZYXI,2024-03-31,0.1034,0.1317,0.8486,-0.9936,0.1366,0.9997,0.9569,0.0693,-0.1207,1.6669
ZYXI,2024-06-30,0.1097,0.1186,-0.1795,-0.6371,-0.2009,-0.0133,-0.4741,0.5304,-0.3476,-1.0715
ZYXI,2024-09-30,0.0010,-0.0142,-0.3068,-0.3372,-0.1675,-0.0276,-0.2824,-0.2045,-0.9928,-0.9823
ZYXI,2024-12-31,-0.0277,-0.0276,NaN,NaN,-0.1083,-0.0485,-0.1110,-0.6049,-1.0229,-0.9832
ZYXI,2025-03-31,-0.4288,-0.5109,NaN,NaN,-0.1621,-0.0874,-0.2742,-6.0788,0.0980,-0.6253


## Profitability Ratios

Profitability ratios measure a company’s ability to convert revenue and assets into profit. They reveal the strength of the firm’s core business model and its capacity to generate returns for shareholders. Metrics like gross margin, operating margin, net margin, ROA, and ROE help distinguish companies that merely grow revenue from those that grow value. These ratios also highlight competitive advantages, pricing power, cost discipline, and long‑term sustainability. In downstream analysis, profitability ratios often serve as leading indicators of strategic success and capital allocation effectiveness.

In [ ]:
DROP TABLE IF EXISTS metrics_profitability;

CREATE TABLE metrics_profitability AS
SELECT
    isq.Ticker,
    isq.ReportDate,
    isq.GrossProfit / NULLIF(isq.Revenue, 0) AS GrossMargin,
    isq.OperatingIncomeLoss / NULLIF(isq.Revenue, 0) AS OperatingMargin,
    isq.NetIncome / NULLIF(isq.Revenue, 0) AS ProfitMargin,
    isq.NetIncome / NULLIF(
        LAG(bsq.TotalAssets, 1) OVER (
            PARTITION BY
                bsq.Ticker
            ORDER BY bsq.ReportDate
        ),
        0
    ) AS ROA,
    isq.NetIncome / NULLIF(
        LAG(bsq.TotalEquity, 1) OVER (
            PARTITION BY
                bsq.Ticker
            ORDER BY bsq.ReportDate
        ),
        0
    ) AS ROE,
    isq.OperatingIncomeLoss * (1 - misc.EffectiveTaxRate) / (
        NULLIF(bsq.TotalEquity, 0) + NULLIF(bsq.ShortTermDebt, 0) + NULLIF(bsq.LongTermDebt, 0) - NULLIF(bsq.CashAndEquiv, 0)
    ) AS ROIC,
    isq.OperatingIncomeLoss / NULLIF(
        LAG(misc.EnterpriseValue, 1) OVER (
            PARTITION BY
                misc.Ticker
            ORDER BY misc.ReportDate
        ),
        0
    ) AS EBITToEV
FROM
    us_income_quarterly isq
    LEFT JOIN us_cashflow_quarterly cfq USING(Ticker, ReportDate)
    LEFT JOIN us_balance_quarterly bsq USING(Ticker, ReportDate)
    LEFT JOIN metrics_misc misc USING(Ticker, ReportDate)

In [ ]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)
\
df = pd.read_sql("SELECT * FROM metrics_profitability", engine)

# -- df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.tail().to_html(index=False))


Ticker,ReportDate,GrossMargin,OperatingMargin,ProfitMargin,ROA,ROE,ROIC,EBITToEV
ZYXI,2024-03-31,0.8002,0.0113,0.0002,0.0001,0.0002,NaN,NaN
ZYXI,2024-06-30,0.8001,0.0446,0.0244,0.0096,0.0368,NaN,NaN
ZYXI,2024-09-30,0.7963,0.0761,0.0477,0.0195,0.0726,NaN,NaN
ZYXI,2024-12-31,0.7829,-0.0122,-0.0134,-0.0049,-0.0172,NaN,NaN
ZYXI,2025-03-31,0.6851,-0.4928,-0.3912,-0.0852,-0.2911,NaN,NaN


## Efficiency (Activity) Ratios

Efficiency ratios assess how effectively a company uses its assets and manages its operations. They highlight the speed and productivity with which resources are converted into revenue. Ratios such as asset turnover, inventory turnover, receivables turnover, and payables turnover illuminate operational discipline, supply‑chain management, and customer‑collection performance. These metrics often uncover bottlenecks or strengths that profitability ratios alone cannot reveal. Efficiency ratios are especially valuable for operational benchmarking and optimization.

In [ ]:
DROP TABLE IF EXISTS metrics_activity;

CREATE TABLE metrics_activity AS
SELECT
    isq.Ticker,
    isq.ReportDate,
    (
        SELECT pr.Date
        FROM us_shareprices_daily pr
        WHERE
            pr.Ticker = isq.Ticker
            AND pr.Date >= isq.ReportDate
        ORDER BY pr.Date ASC
        LIMIT 1
    ) AS PricesDate,
    (
        SELECT pr.Close
        FROM us_shareprices_daily pr
        WHERE
            pr.Ticker = isq.Ticker
            AND pr.Date >= isq.ReportDate
        ORDER BY pr.Date ASC
        LIMIT 1
    ) / NULLIF(isq.NetIncome, 0) AS PE,
    (
        SELECT pr.Close
        FROM us_shareprices_daily pr
        WHERE
            pr.Ticker = isq.Ticker
            AND pr.Date >= isq.ReportDate
        ORDER BY pr.Date ASC
        LIMIT 1
    ) / NULLIF(isq.Revenue, 0) AS PS,
    (
        SELECT pr.Close
        FROM us_shareprices_daily pr
        WHERE
            pr.Ticker = isq.Ticker
            AND pr.Date >= isq.ReportDate
        ORDER BY pr.Date ASC
        LIMIT 1
    ) / NULLIF(bsq.TotalEquity, 0) AS PB
FROM
    us_income_quarterly isq
    LEFT JOIN us_cashflow_quarterly cfq USING(Ticker, ReportDate)
    LEFT JOIN us_balance_quarterly bsq USING(Ticker, ReportDate)
    LEFT JOIN metrics_misc misc USING(Ticker, ReportDate)

In [ ]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)


df = pd.read_sql("SELECT * FROM metrics_activity", engine)

HTML(df.tail().to_html(index=False))

## Value Ratios

Valuation ratios connect fundamentals to market expectations. They measure how investors price a company relative to its earnings, cash flow, book value, or sales. Ratios like P/E, P/B, P/S, and EV/EBITDA help determine whether a stock appears undervalued, fairly valued, or overvalued. These metrics reflect sentiment, growth expectations, and perceived risk. In downstream analysis, valuation ratios bridge the gap between internal performance and external market behavior, enabling more informed investment decisions.

In [ ]:
DROP TABLE IF EXISTS metrics_value;

CREATE TABLE metrics_value AS
WITH
    ratios_price_date AS (
        SELECT isq.Ticker, isq.ReportDate, (
                SELECT pr.Date
                FROM us_shareprices_daily pr
                WHERE
                    pr.Ticker = isq.Ticker
                    AND pr.Date >= isq.ReportDate
                ORDER BY pr.Date ASC
                LIMIT 1
            ) AS PricesDate
        FROM
            us_shareprices_daily pr
            LEFT JOIN us_income_quarterly isq ON pr.Ticker = isq.Ticker
            AND isq.ReportDate = pr.Date
    )
SELECT isq.Ticker, isq.ReportDate, rpd.PricesDate, (
        SELECT pr.Close
        FROM us_shareprices_daily pr
        WHERE
            pr.Ticker = isq.Ticker
            AND pr.Date = rpd.PricesDate
    ) / 
    CASE 
        WHEN isq.NetIncome IS NULL OR isq.NetIncome < 0 THEN NULL
        ELSE (
            SELECT misc.MarketCap 
            FROM us_shareprices_daily pr
            WHERE
                pr.Ticker = isq.Ticker
                AND misc.ReportDate = rpd.PricesDate
        ) / isq.NetIncome
    END AS PE, (
        SELECT misc.MarketCap
        FROM metrics_misc misc
        WHERE
            misc.Ticker = isq.Ticker
            AND misc.ReportDate = rpd.PricesDate
    ) / NULLIF(isq.Revenue, 0) AS PS, (
        SELECT misc.MarketCap
        FROM metrics_misc misc
        WHERE
            misc.Ticker = isq.Ticker
            AND misc.ReportDate = rpd.PricesDate
    ) / NULLIF(bsq.TotalEquity, 0) AS PB
FROM
    us_income_quarterly isq
    LEFT JOIN us_cashflow_quarterly cfq USING(Ticker, ReportDate)
    LEFT JOIN us_balance_quarterly bsq USING(Ticker, ReportDate)
    LEFT JOIN metrics_misc misc USING(Ticker, ReportDate)
    LEFT JOIN ratios_price_date rpd USING(Ticker, ReportDate)

In [ ]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)


df = pd.read_sql("SELECT * FROM metrics_value", engine)

HTML(df.tail().to_html(index=False))

Ticker,ReportDate,PE,PS,PB
ZUO,2021-04-30,-9.200000e-07,2.000000e-07,11.462709
ZYME,2020-12-31,-1.250000e-06,3.020000e-06,5.905741
ZYNE,2020-12-31,-3.400000e-07,NaN,1.610323
ZYNE,2021-03-31,-5.800000e-07,NaN,1.946462
ZYNE,2022-12-31,NaN,NaN,NaN


## Liquidity Ratios

Liquidity ratios evaluate a company’s ability to meet short‑term obligations using its most liquid assets. They answer a fundamental question: Can the company pay its bills without stress? Ratios such as the current ratio, quick ratio, and cash ratio help identify operational resilience and working‑capital efficiency. Strong liquidity supports stability during downturns, while weak liquidity can signal cash‑flow strain or reliance on external financing. These ratios are essential for risk assessment, especially in volatile industries or early‑stage companies.

In [ ]:

DROP TABLE IF EXISTS metrics_liquidity;

CREATE TABLE metrics_liquidity AS (
SELECT
    Ticker,
    ReportDate,
    CashAndEquiv / NULLIF(TotalCurrentAssets, 0) AS CashRatio,
    (TotalCurrentAssets - Inventories) / NULLIF(TotalCurrentAssets, 0) AS QuickRatio,
    (TotalCurrentAssets / NULLIF(TotalCurrentLiabilities, 0)) AS CurrentRatio
FROM us_balance_quarterly
)

In [65]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)


df = pd.read_sql("SELECT * FROM metrics_liquidity", engine)

HTML(df.tail().to_html(index=False))

Ticker,ReportDate,CashRatio,QuickRatio,CurrentRatio
ZYXI,2024-03-31,0.4214,0.8016,3.5849
ZYXI,2024-06-30,0.4169,0.7923,4.0638
ZYXI,2024-09-30,0.4803,0.7995,3.9435
ZYXI,2024-12-31,0.5272,0.8149,4.4580
ZYXI,2025-03-31,0.4228,0.7601,3.4574


## Solvency Ratios

Solvency ratios measure long‑term financial stability and the degree to which a company relies on debt to finance operations. Metrics like debt‑to‑equity, interest coverage, and debt‑to‑assets reveal leverage levels and the firm’s ability to service debt over time. High leverage can amplify returns but also magnify risk. Solvency ratios help analysts understand capital structure, creditworthiness, and exposure to macroeconomic conditions. They are critical for evaluating long‑term viability and strategic flexibility.

In [ ]:
DROP TABLE IF EXISTS metrics_solvency;

CREATE TABLE metrics_solvency AS (
SELECT
    bsq.Ticker,
    bsq.ReportDate,
    -- Debt-to-Equity: total debt relative to shareholder equity
    (misc.TotalDebt / NULLIF(bsq.TotalEquity, 0)) AS DebtToEquity,
    -- Debt-to-Assets: leverage relative to total assets
    (bsq.TotalLiabilities / NULLIF(bsq.TotalAssets, 0)) AS DebtRatio,
    -- Interest Coverage: ability to cover interest expense with EBIT
    CASE WHEN isq.InterestExpenseNet > 0  OR isq.OperatingIncomeLoss < 0 THEN NULL
    ELSE (isq.OperatingIncomeLoss / NULLIF(ABS(isq.InterestExpenseNet), 0))
    END AS InterestCoverageRatio
FROM us_balance_quarterly bsq
-- LEFT JOIN us_income_quarterly isq USING (Ticker, ReportDate)
LEFT JOIN metrics_misc misc USING(Ticker, ReportDate)
LEFT JOIN us_income_quarterly isq USING(Ticker, ReportDate)
);


## Cash Flow Quality Ratios

Cash flow quality reflects how reliably a company turns its accounting earnings into actual cash, offering a deeper look into the sustainability of its operations beyond reported profit figures; metrics such as Earnings Quality, FCF Margin, and FCF Conversion reveal whether earnings are supported by cash generation or distorted by accruals, capital intensity, or working‑capital swings, and across the companies shown, ZUO demonstrates weak cash flow quality with negative earnings support and limited free‑cash generation, ZYME shows unusually strong earnings quality but poor FCF margin - often a sign of heavy investment or non‑recurring cash movements - and ZYNE consistently converts earnings into cash effectively despite missing FCF margin data, suggesting disciplined cash management; overall, cash flow quality acts as a critical lens for distinguishing durable performance from accounting noise, identifying firms capable of self‑funding growth, and assessing resilience when paired with profitability, efficiency, liquidity, and solvency ratios.

In [ ]:

DROP TABLE IF EXISTS metrics_cashflow_quality;

CREATE TABLE metrics_cashflow_quality AS (

SELECT
    isq.Ticker,
    isq.ReportDate,
    cfq.NetCashFromOperatingActivities / NULLIF(isq.NetIncome, 0) AS EarningsQuality,
    misc.FreeCashFlow / NULLIF(isq.Revenue, 0) AS FCFMargin,
    misc.FreeCashFlow / NULLIF(isq.NetIncome, 0) AS FCFConversion
FROM us_cashflow_quarterly cfq
LEFT JOIN us_income_quarterly isq USING(Ticker, ReportDate)
LEFT JOIN us_balance_quarterly bsq USING(Ticker, ReportDate)
LEFT JOIN metrics_misc misc USING(Ticker, ReportDate)
)

In [68]:
# HideCell

from IPython.display import HTML
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}"
)


df = pd.read_sql("SELECT * FROM metrics_cashflow_quality", engine)

HTML(df.tail().to_html(index=False))

Ticker,ReportDate,EarningsQuality,FCFMargin,FCFConversion
ZUO,2021-04-30,-0.5804,0.1032,-0.4692
ZYME,2020-12-31,1.2642,-3.1242,1.2912
ZYNE,2020-12-31,0.7264,NaN,0.7329
ZYNE,2021-03-31,1.0416,NaN,1.0416
ZYNE,2022-12-31,0.7093,NaN,0.7102


# Conclusion

By converting raw fundamentals into a structured suite of ratios, this project creates a powerful analytical engine for understanding corporate performance. Each ratio group provides a different lens - profitability shows value creation, liquidity shows short‑term stability, solvency shows long‑term risk, efficiency shows operational execution, and valuation shows market perception. Together, they form a holistic framework that supports modeling, screening, forecasting, and strategic insights. This ratio‑driven foundation ensures that downstream analysis is not only data‑rich but also economically meaningful.